[SRVKE-1278 + SRVKE-1327] : Add kafka sink security config docs (#50970)
openshift/openshift-docs: modules/serverless-kafka-sink-security-config.adoc

_Simple Authentication and Security Layer_ (SASL) is used by Apache Kafka for authentication. If you use SASL authentication on your cluster, users must provide credentials to Knative for communicating with the Kafka cluster, otherwise events cannot be produced or consumed.

.Prerequisites

* The {ServerlessOperatorName}, Knative Eventing, and the `KnativeKafka` custom resources (CRs) are installed on your {product-title} cluster.
* Kafka sink is enabled in the `KnativeKafka` CR.
* You have created a project or have access to a project with the appropriate roles and permissions to create applications and other workloads in {product-title}.
* You have a Kafka cluster CA certificate stored as a `.pem` file.
* You have a Kafka cluster client certificate and a key stored as `.pem` files.
* You have installed the OpenShift (`oc`) CLI.

.Procedure

. Create the certificate files as a secret in the same namespace as your `KafkaSink` object:
+
[IMPORTANT]
====
Certificates and keys must be in PEM format.
====

** For authentication using SASL without encryption:
+
[source,terminal]
----
$ oc create secret -n <namespace> generic <secret_name> \
  --from-literal=protocol=SASL_PLAINTEXT \

>   --from-literal=sasl.mechanism=<sasl_mechanism> \

Reviewer: I have noticed we describe what these are. (we do at least list the options in kafka source docs, in prereqs: 
```You have chosen the SASL mechanism to use, for example, PLAIN, SCRAM-SHA-256, or SCRAM-SHA-512.```

I think the important hint for those is that these need to be uppercase, so just listing these options anywhere in the doc would help.

  --from-literal=user=<username> \
  --from-literal=password=<password>
----

** For authentication using SASL and encryption using TLS:
+
[source,terminal]
----
$ oc create secret -n <namespace> generic <secret_name> \
  --from-literal=protocol=SASL_SSL \
  --from-literal=sasl.mechanism=<sasl_mechanism> \

>   --from-file=ca.crt=caroot.pem \ <1>

Reviewer: ```suggestion
  --from-file=<my_caroot.pem_file_path> \ <1>
```

for consistency with below

Author: We actually use this format for the other docs as well https://50970--docspreview.netlify.app/openshift-enterprise/latest/serverless/admin_guide/serverless-kafka-admin.html#serverless-kafka-broker-sasl-default-config_serverless-kafka-admin so I think if we change it here we'd need to change it across the board, and that it should be something more specific / easy to understand than `<my_caroot.pem_file_path>`?

Author: Maybe something like `<ca_root>`?

  --from-literal=user=<username> \
  --from-literal=password=<password>
----
<1> The `ca.crt` can be omitted to use the system's root CA set. For example, if you are using a public cloud managed Kafka service such as Red Hat OpenShift Streams for Apache Kafka.

** For encryption using TLS without client authentication:
+
[source,terminal]
----
$ oc create secret -n <namespace> generic <secret_name> \
  --from-literal=protocol=SSL \
  --from-file=ca.crt=<my_caroot.pem_file_path> \ <1>
  --from-literal=user.skip=true
----

> <1> The `ca.crt` can be omitted to use the system's root CA set. For example, if you are using a public cloud managed Kafka service such as Red Hat OpenShift Streams for Apache Kafka.

Reviewer: so, this bit doesn't work, the docs is correct though, the bug is in the product. I have opened https://issues.redhat.com/browse/SRVKE-1328

should we remove it from docs, or just leave it there and perhaps release-note the issue?

Author: Does it work for upstream or should we also remove it there or open a bug at the source?

Reviewer: There's an open issue in upstream https://github.com/knative-sandbox/eventing-kafka-broker/issues/2708

// Module is included in the following assemblies:
//
// * serverless/admin_guide/serverless-kafka-admin.adoc

:_content-type: PROCEDURE
[id="serverless-kafka-sink-security-config_{context}"]
= Configuring security for Kafka sinks

> You can configure security features, such as encryption and authentication, for Kafka sinks. You can do this by creating certificates as secrets, and then referencing these secrets from the `KafkaSink` object that you want to configure security for.

Reviewer: I think this could be a bit misleading. The TLS and SASL configuration here is only for encryption and authentication that KafkaSink uses to connect to Kafka.

It doesn't configure security for the KafkaSink itself (anybody who can connect to the kafkasink's URL can still POST any cloudevents to the kafkasink)

Reviewer 2: I think it is a valid concern perhaps the wording in https://github.com/openshift/openshift-docs/pull/50877/files seems concise enough and less misleading.

```
= Configuring SASL authentication for Kafka sinks

_Simple Authentication and Security Layer_ (SASL) is used by Apache Kafka for authentication. If you use SASL authentication on your cluster, users must provide credentials to Knative for communicating with the Kafka cluster, otherwise events cannot be produced or consumed.
```

Reviewer 2: > (Any idea about a better wording or maybe explicitly say something about having the sink still open to anything?)

We can perhaps capture an issue for the open to anything problem and rework every addressable that we ship to mention the open to anything problem and how to address it

Author: Removing the paragraph
> You can configure security features, such as encryption and authentication, for Kafka sinks. You can do this by creating certificates as secrets, and then referencing these secrets from the `KafkaSink` object that you want to configure security for.

and leaving the part about SASL and TLS. Does this solve the issue?
@pierDipi can you open another issue please if you're saying that you think that's required and we'll need a follow up?

** For encryption using TLS without client authentication:
+
[source,terminal]
----
$ oc create secret -n <namespace> generic <secret_name> \
  --from-literal=protocol=SSL \
  --from-file=ca.crt=<my_caroot.pem_file_path> \ <1>
  --from-literal=user.skip=true
----
<1> The `ca.crt` can be omitted to use the system's root CA set. For example, if you are using a public cloud managed Kafka service such as Red Hat OpenShift Streams for Apache Kafka.

** For authentication and encryption using TLS:
+
[source,terminal]
----
$ oc create secret -n <namespace> generic <secret_name> \
  --from-literal=protocol=SSL \
  --from-file=ca.crt=<my_caroot.pem_file_path> \ <1>
  --from-file=user.crt=<my_cert.pem_file_path> \
  --from-file=user.key=<my_key.pem_file_path>
----
<1> The `ca.crt` can be omitted to use the system's root CA set. For example, if you are using a public cloud managed Kafka service such as Red Hat OpenShift Streams for Apache Kafka.

. Create or modify a `KafkaSink` object and add a reference to your secret in the `auth` spec:
+
[source,yaml]
----
apiVersion: eventing.knative.dev/v1alpha1
kind: KafkaSink
metadata:
   name: <sink_name>

>    namespace: default

Reviewer 2: for consistency of the doc and the original sentence "Create the certificate files as a secret in the same namespace as your `KafkaSink` object:"
```suggestion
   namespace: <namespace>
```

_Transport Layer Security_ (TLS) is used by Apache Kafka clients and servers to encrypt traffic between Knative and Kafka, as well as for authentication. TLS is the only supported method of traffic encryption for Knative Kafka.

_Simple Authentication and Security Layer_ (SASL) is used by Apache Kafka for authentication. If you use SASL authentication on your cluster, users must provide credentials to Knative for communicating with the Kafka cluster, otherwise events cannot be produced or consumed.

.Prerequisites

* The {ServerlessOperatorName}, Knative Eventing, and the `KnativeKafka` custom resources (CRs) are installed on your {product-title} cluster.
* Kafka sink is enabled in the `KnativeKafka` CR.
* You have created a project or have access to a project with the appropriate roles and permissions to create applications and other workloads in {product-title}.
* You have a Kafka cluster CA certificate stored as a `.pem` file.
* You have a Kafka cluster client certificate and a key stored as `.pem` files.
* You have installed the OpenShift CLI (`oc`).

.Procedure

> . Create the certificate files as a secret in the `knative-eventing` namespace:
> +
> [IMPORTANT]
> ====
> You must create the secret in the same namespace as your `KafkaSink` object.

Reviewer 2: I would remove the "Create the certificate files as a secret in the `knative-eventing` namespace:" since "You must create the secret in the same namespace as your `KafkaSink` object" is basically negating it.

```suggestion
[IMPORTANT]
====
You must create the secret in the same namespace as your `KafkaSink` object.
```

Certificates and keys must be in PEM format.

> The `ca.crt` can be omitted to enable fallback and use the system's root CA set.

Reviewer 2: A bit convoluted, what about something like:
```suggestion
The `ca.crt` can be omitted to fall back to use the system's root CA set.
```

Reviewer 2: In addition, I would add an example when this is used:
```suggestion
The `ca.crt` can be omitted to fall back to using the system's root CA set when using a public cloud managed Kafka service like Red Hat OpenShift Streams for Apache Kafka.
```

+
[source,terminal]
----
$ oc create secret -n <namespace> generic <secret_name> \
  --from-literal=protocol=SASL_PLAINTEXT \
  --from-literal=sasl.mechanism=<sasl_mechanism> \
  --from-literal=user=<username> \
  --from-literal=password=<password>
----

** For authentication using SASL and encryption using TLS:
+
[source,terminal]
----
$ oc create secret -n <namespace> generic <secret_name> \
  --from-literal=protocol=SASL_SSL \
  --from-literal=sasl.mechanism=<sasl_mechanism> \
  --from-file=ca.crt=caroot.pem \
  --from-literal=user=<username> \
  --from-literal=password=<password>
----

** For encryption using TLS without client authentication:
+
[source,terminal]
----
$ oc create secret -n <namespace> generic <secret_name> \
  --from-literal=protocol=SSL \
  --from-file=ca.crt=<my_caroot.pem_file_path> \
  --from-literal=user.skip=true
----

** For authentication and encryption using TLS:
+
[source,terminal]
----
$ oc create secret -n <namespace> generic <secret_name> \
  --from-literal=protocol=SSL \
  --from-file=ca.crt=<my_caroot.pem_file_path> \
  --from-file=user.crt=<my_cert.pem_file_path> \
  --from-file=user.key=<my_key.pem_file_path>
----

> . Create or modify a `KnativeSink` object and add a reference to your secret in the `auth` spec:

Reviewer 2: ```suggestion
. Create or modify a `KafkaSink` object and add a reference to your secret in the `auth` spec:
```

+
[source,yaml]
----
apiVersion: eventing.knative.dev/v1alpha1
kind: KafkaSink
metadata:
   name: <sink_name>
   namespace: default
spec:
...
   auth:
     secret:
       ref:
         name: <secret_name>
...
----

> . Apply the `KnativeSink` object:

Reviewer 2: ```suggestion
. Apply the `KafkaSink` object:
```

Author: Oops!

// Module is included in the following assemblies:
//
// * serverless/admin_guide/serverless-kafka-admin.adoc

:_content-type: PROCEDURE
[id="serverless-kafka-sink-security-config_{context}"]
= Configuring security for Kafka sinks

> 

Reviewer 3: Need [role="_abstract"] and abstract for the module per https://raw.githubusercontent.com/redhat-documentation/modular-docs/master/modular-docs-manual/files/TEMPLATE_PROCEDURE_doing-one-procedure.adoc

Author: The last time I asked @kalexand-rh about using `[role="_abstract"]` the guidance AFAIK was that we don't need to use it, so it's currently not used anywhere in our docs. Kathryn, should I open a Jira for serverless to update all the abstracts to use this, or is it still not required?

Reviewer 3: This might be just for OpenShift (I'm new to peer reviewing) :)

Reviewer 4: `[role="_abstract"]` is only for PV2 - it's been removed from the Jupiter guidelines, which OpenShift uses. RHACS is the only content in the OCP repo that requires that role. (The additional resource one is required.)

_Transport Layer Security_ (TLS) is used by Apache Kafka clients and servers to encrypt traffic between Knative and Kafka, as well as for authentication. TLS is the only supported method of traffic encryption for Knative Kafka.

> _Simple Authentication and Security Layer_ (SASL) is used by Apache Kafka for authentication. If you use SASL authentication on your cluster, users must provide credentials to Knative for communicating with the Kafka cluster, otherwise events cannot be produced or consumed.

Reviewer 3: I think you need a semicolon before "otherwise" (independent clause - see IBM SG)

Author: Updated across docs, thanks!

.Prerequisites

* The {ServerlessOperatorName}, Knative Eventing, and the `KnativeKafka` custom resources (CRs) are installed on your {product-title} cluster.
* Kafka sink is enabled in the `KnativeKafka` CR.
* You have created a project or have access to a project with the appropriate roles and permissions to create applications and other workloads in {product-title}.
* You have a Kafka cluster CA certificate stored as a `.pem` file.
* You have a Kafka cluster client certificate and a key stored as `.pem` files.
* You have installed the OpenShift (`oc`) CLI.

> * You have chosen the SASL mechanism to use, for example, `PLAIN`, `SCRAM-SHA-256`, or `SCRAM-SHA-512`.

Reviewer 3: Suggestion: remove "to use" and add semicolon ("You have chosen the SASL mechanism; for example, ...."

Author: I think this is OK as it is

.Procedure

. Create the certificate files as a secret in the same namespace as your `KafkaSink` object:
+
[IMPORTANT]
====
Certificates and keys must be in PEM format.
====

** For authentication using SASL without encryption:
+
[source,terminal]
----
$ oc create secret -n <namespace> generic <secret_name> \
  --from-literal=protocol=SASL_PLAINTEXT \
  --from-literal=sasl.mechanism=<sasl_mechanism> \
  --from-literal=user=<username> \
  --from-literal=password=<password>
----

** For authentication using SASL and encryption using TLS:
+
[source,terminal]
----
$ oc create secret -n <namespace> generic <secret_name> \
  --from-literal=protocol=SASL_SSL \
  --from-literal=sasl.mechanism=<sasl_mechanism> \
  --from-file=ca.crt=<my_caroot.pem_file_path> \ <1>
  --from-literal=user=<username> \
  --from-literal=password=<password>
----

> <1> The `ca.crt` can be omitted to use the system's root CA set. For example, if you are using a public cloud managed Kafka service such as Red Hat OpenShift Streams for Apache Kafka.

Reviewer 3: Suggestion: The second sentence seems incomplete - maybe add wording? "For example, use the system's root CA set if you are using..." or "For example, omit the `ca.crt` if you are using..." Or change the period after set to a semicolon (but that makes the sentence long).

** For authentication and encryption using TLS:
+
[source,terminal]
----
$ oc create secret -n <namespace> generic <secret_name> \
  --from-literal=protocol=SSL \
  --from-file=ca.crt=<my_caroot.pem_file_path> \ <1>
  --from-file=user.crt=<my_cert.pem_file_path> \
  --from-file=user.key=<my_key.pem_file_path>
----

> <1> The `ca.crt` can be omitted to use the system's root CA set. For example, if you are using a public cloud managed Kafka service such as Red Hat OpenShift Streams for Apache Kafka.

Reviewer 3: same suggestion as above
